File: sdk/typescript/src/dpop.ts
/**
 * DPoP proof JWT generation per RFC 9449.
 *
 * Holds an ECDSA P-256 keypair and emits DPoP proof JWTs bound to an HTTP
 * method, URL, and optionally an access token (via the `ath` claim).
 *
 * @example
 * ```ts
 * const prover = await DPoPProver.generate();
 * const proof = await prover.createProof({ method: "POST", url: "https://auth.example/token" });
 * ```
 */

import {
  exportPKCS8,
  exportJWK,
  generateKeyPair,
  importPKCS8,
  type KeyLike,
  type JWK,
} from "jose";
import { SignJWT } from "jose";
import { DPoPError } from "./errors.js";

/** Options for {@link DPoPProver.createProof}. */
export interface CreateProofOptions {
  /** HTTP method (e.g. `"POST"`). Normalized to uppercase. */
  method: string;
  /** Target URL (without query/fragment per RFC 9449 §4.2). */
  url: string;
  /**
   * Optional server-provided DPoP nonce (from `DPoP-Nonce` response header).
   * Included in the proof as the `nonce` claim.
   */
  nonce?: string;
  /**
   * If provided, the proof binds to the token via the `ath` claim
   * (`base64url(sha256(accessToken))`).
   */
  accessToken?: string;
  /** Override `iat` (for testing only). */
  iat?: number;
  /** Override `jti` (for testing only). */
  jti?: string;
}

/** @internal */
interface EcPublicJwk {
  kty: "EC";
  crv: "P-256";
  x: string;
  y: string;
}

/**
 * RFC 9449 DPoP prover.
 *
 * Typically constructed via {@link DPoPProver.generate} (fresh P-256 key)
 * or {@link DPoPProver.fromPem} (load an existing PEM-encoded private key).
 */
export class DPoPProver {
  private readonly _privateKey: KeyLike;
  private readonly _publicJwk: EcPublicJwk;
  private readonly _thumbprint: string;

  private constructor(
    privateKey: KeyLike,
    publicJwk: EcPublicJwk,
    thumbprint: string
  ) {
    this._privateKey = privateKey;
    this._publicJwk = publicJwk;
    this._thumbprint = thumbprint;
  }

  // ------------------------------------------------------------------
  // Constructors
  // ------------------------------------------------------------------

  /**
   * Generate a fresh ECDSA P-256 keypair and return a new `DPoPProver`.
   */
  static async generate(): Promise<DPoPProver> {
    const { privateKey } = await generateKeyPair("ES256", {
      crv: "P-256",
      extractable: true,
    });
    return DPoPProver._fromPrivateKey(privateKey);
  }

  /**
   * Load a PEM-encoded PKCS#8 private key and return a new `DPoPProver`.
   *
   * @param pem  PKCS#8 PEM string (or `Buffer`).
   * @throws {@link DPoPError} if the key is not a P-256 EC key.
   */
  static async fromPem(pem: string | Buffer): Promise<DPoPProver> {
    const pemStr = Buffer.isBuffer(pem) ? pem.toString("utf-8") : pem;
    let privateKey: KeyLike;
    try {
      privateKey = await importPKCS8(pemStr, "ES256");
    } catch (err) {
      throw new DPoPError(
        `Failed to load private key: ${err instanceof Error ? err.message : String(err)}`
      );
    }
    return DPoPProver._fromPrivateKey(privateKey);
  }

  /** @internal */
  private static async _fromPrivateKey(privateKey: KeyLike): Promise<DPoPProver> {
    let rawJwk: JWK;
    try {
      rawJwk = await exportJWK(privateKey);
    } catch (err) {
      throw new DPoPError(
        `Failed to export JWK: ${err instanceof Error ? err.message : String(err)}`
      );
    }

    if (rawJwk.kty !== "EC" || rawJwk.crv !== "P-256") {
      throw new DPoPError("Private key must be ECDSA P-256 (secp256r1)");
    }
    if (typeof rawJwk.x !== "string" || typeof rawJwk.y !== "string") {
      throw new DPoPError("Private key missing EC public coordinates x/y");
    }

    const publicJwk: EcPublicJwk = {
      kty: "EC",
      crv: "P-256",
      x: rawJwk.x,
      y: rawJwk.y,
    };

    const thumbprint = await _jwkThumbprint(publicJwk);
    return new DPoPProver(privateKey, publicJwk, thumbprint);
  }

  // ------------------------------------------------------------------
  // Accessors
  // ------------------------------------------------------------------

  /**
   * The RFC 7638 JWK thumbprint (SHA-256, base64url) bound to this keypair.
   * Include as `dpop_jkt` in authorization requests that support it.
   */
  get jkt(): string {
    return this._thumbprint;
  }

  /**
   * The public key as a JWK (copy — no private material).
   */
  get publicJwk(): EcPublicJwk {
    return { ...this._publicJwk };
  }

  /**
   * Export the private key as an unencrypted PKCS#8 PEM string.
   *
   * Store this securely; it is sufficient to reconstruct the prover via
   * {@link DPoPProver.fromPem}.
   */
  async privateKeyPem(): Promise<string> {
    return exportPKCS8(this._privateKey);
  }

  // ------------------------------------------------------------------
  // Proof emission
  // ------------------------------------------------------------------

  /**
   * Build and sign a DPoP proof JWT.
   *
   * @param opts  Proof options including `method` and `url` (required).
   * @returns Compact serialised DPoP JWT string.
   * @throws {@link DPoPError} if `method` or `url` is empty.
   */
  async createProof(opts: CreateProofOptions): Promise<string> {
    const { method, nonce, accessToken, iat, jti } = opts;
    let { url } = opts;

    if (!method) throw new DPoPError("method is required");
    if (!url) throw new DPoPError("url is required");

    // RFC 9449 §4.2: The HTU claim MUST NOT contain any query or fragment.
    if (url.includes("?")) url = url.split("?")[0]!;
    if (url.includes("#")) url = url.split("#")[0]!;

    const claims: Record<string, unknown> = {
      jti: jti ?? _randomJti(),
      htm: method.toUpperCase(),
      htu: url,
      iat: iat ?? Math.floor(Date.now() / 1000),
    };

    if (nonce !== undefined) {
      claims["nonce"] = nonce;
    }
    if (accessToken !== undefined) {
      claims["ath"] = await _ath(accessToken);
    }

    try {
      return await new SignJWT(claims)
        .setProtectedHeader({
          typ: "dpop+jwt",
          alg: "ES256",
          jwk: this._publicJwk,
        })
        .sign(this._privateKey);
    } catch (err) {
      throw new DPoPError(
        `Failed to sign DPoP proof: ${err instanceof Error ? err.message : String(err)}`
      );
    }
  }
}

// ------------------------------------------------------------------
// Internal helpers
// ------------------------------------------------------------------

/** Compute `base64url(sha256(accessToken))` for the `ath` claim. */
async function _ath(accessToken: string): Promise<string> {
  const encoder = new TextEncoder();
  const digest = await crypto.subtle.digest(
    "SHA-256",
    encoder.encode(accessToken)
  );
  return _b64url(new Uint8Array(digest));
}

/** RFC 7638 JWK thumbprint for a P-256 EC public key. */
async function _jwkThumbprint(jwk: EcPublicJwk): Promise<string> {
  // Canonical form: only required members, lexicographic key order.
  const canonical = JSON.stringify({
    crv: jwk.crv,
    kty: jwk.kty,
    x: jwk.x,
    y: jwk.y,
  });
  const encoder = new TextEncoder();
  const digest = await crypto.subtle.digest("SHA-256", encoder.encode(canonical));
  return _b64url(new Uint8Array(digest));
}

/** base64url-encode a byte array without padding. */
function _b64url(bytes: Uint8Array): string {
  let binary = "";
  for (let i = 0; i < bytes.byteLength; i++) {
    binary += String.fromCharCode(bytes[i]);
  }
  return btoa(binary).replace(/\+/g, "-").replace(/\//g, "_").replace(/=/g, "");
}

/** Generate a random base64url JTI (~128 bits of entropy). */
function _randomJti(): string {
  const bytes = new Uint8Array(16);
  crypto.getRandomValues(bytes);
  return _b64url(bytes);
}

File: sdk/python/shark_auth/dpop.py
"""DPoP proof JWT generation per RFC 9449.

Holds an ECDSA P-256 keypair and emits DPoP proof JWTs bound to an HTTP
method, URL, and optionally an access token (via the ``ath`` claim).
"""

from __future__ import annotations

import base64
import hashlib
import json
import secrets
import time
from dataclasses import dataclass
from typing import Optional

import jwt
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import ec

from .errors import DPoPError


def _b64url(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode("ascii")


def _int_to_b64url(value: int, length: int) -> str:
    return _b64url(value.to_bytes(length, "big"))


@dataclass
class DPoPProver:
    """RFC 9449 DPoP prover.

    Typically constructed via :meth:`generate` (fresh P-256 key) or
    :meth:`from_pem` (load an existing PEM-encoded private key).
    """

    _private_key: ec.EllipticCurvePrivateKey
    _public_jwk: dict
    _thumbprint: str

    # ------------------------------------------------------------------
    # Constructors
    # ------------------------------------------------------------------
    @classmethod
    def generate(cls) -> "DPoPProver":
        """Generate a fresh ECDSA P-256 keypair."""
        key = ec.generate_private_key(ec.SECP256R1())
        return cls._from_private_key(key)

    @classmethod
    def from_pem(cls, pem: bytes, password: Optional[bytes] = None) -> "DPoPProver":
        """Load a PEM-encoded private key."""
        try:
            key = serialization.load_pem_private_key(pem, password=password)
        except Exception as exc:  # pragma: no cover - re-wrap
            raise DPoPError(f"failed to load private key: {exc}") from exc
        if not isinstance(key, ec.EllipticCurvePrivateKey) or not isinstance(
            key.curve, ec.SECP256R1
        ):
            raise DPoPError("private key must be ECDSA P-256 (secp256r1)")
        return cls._from_private_key(key)

    @classmethod
    def _from_private_key(cls, key: ec.EllipticCurvePrivateKey) -> "DPoPProver":
        numbers = key.public_key().public_numbers()
        x = _int_to_b64url(numbers.x, 32)
        y = _int_to_b64url(numbers.y, 32)
        jwk = {"kty": "EC", "crv": "P-256", "x": x, "y": y}
        thumbprint = _jwk_thumbprint(jwk)
        return cls(_private_key=key, _public_jwk=jwk, _thumbprint=thumbprint)

    # ------------------------------------------------------------------
    # Accessors
    # ------------------------------------------------------------------
    def private_key_pem(self) -> bytes:
        """Return the private key as unencrypted PEM (PKCS#8)."""
        return self._private_key.private_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PrivateFormat.PKCS8,
            encryption_algorithm=serialization.NoEncryption(),
        )

    @property
    def public_jwk(self) -> dict:
        """Return the public key as a JWK (dict)."""
        return dict(self._public_jwk)

    @property
    def jkt(self) -> str:
        """Return the RFC 7638 JWK thumbprint (SHA-256, base64url)."""
        return self._thumbprint

    # ------------------------------------------------------------------
    # Proof emission
    # ------------------------------------------------------------------
    def make_proof(
        self,
        htm: str,
        htu: str,
        *,
        nonce: Optional[str] = None,
        access_token: Optional[str] = None,
        iat: Optional[int] = None,
        jti: Optional[str] = None,
    ) -> str:
        """Build and sign a DPoP proof JWT.

        Parameters
        ----------
        htm
            HTTP method (e.g. ``"POST"``). Case-insensitive; normalized to upper.
        htu
            Target URL (without query / fragment per RFC 9449 §4.2).
        nonce
            Optional DPoP nonce advertised by the server in a prior
            ``DPoP-Nonce`` response header.
        access_token
            If set, the proof binds to the token via the ``ath`` claim
            (``base64url(sha256(access_token))``).
        iat, jti
            Optional overrides for testing. In production, leave unset.
        """
        if not htm or not htu:
            raise DPoPError("htm and htu are required")

        payload = {
            "jti": jti or _b64url(secrets.token_bytes(16)),
            "htm": htm.upper(),
            "htu": htu,
            "iat": iat if iat is not None else int(time.time()),
        }
        if nonce is not None:
            payload["nonce"] = nonce
        if access_token is not None:
            digest = hashlib.sha256(access_token.encode("ascii")).digest()
            payload["ath"] = _b64url(digest)

        headers = {"typ": "dpop+jwt", "alg": "ES256", "jwk": self._public_jwk}
        pem = self.private_key_pem()
        try:
            return jwt.encode(payload, pem, algorithm="ES256", headers=headers)
        except Exception as exc:  # pragma: no cover
            raise DPoPError(f"failed to sign proof: {exc}") from exc


def _jwk_thumbprint(jwk: dict) -> str:
    """RFC 7638 JWK thumbprint for an EC JWK."""
    # Canonical form: lexicographic keys, no whitespace.
    canonical = {"crv": jwk["crv"], "kty": jwk["kty"], "x": jwk["x"], "y": jwk["y"]}
    serialized = json.dumps(canonical, separators=(",", ":"), sort_keys=True).encode(
        "utf-8"
    )
    return _b64url(hashlib.sha256(serialized).digest())
